src/Interpreters/AsynchronousInsertQueue.h

```diff
@@ -80,62 +71,40 @@ class AsynchronousInsertQueue : public WithContext
 
     using InsertDataPtr = std::unique_ptr<InsertData>;
 
-    /// A separate container, that holds a data and a mutex for it.
-    /// When it's needed to process current chunk of data, it can be moved for processing
-    /// and new data can be recreated without holding a lock during processing.
     struct Container
     {
-        std::mutex mutex;
+        InsertQuery key;
         InsertDataPtr data;
     };
 
-    using Queue = std::unordered_map<InsertQuery, std::shared_ptr<Container>, InsertQuery::Hash>;
-    using QueueIterator = Queue::iterator;
     /// Ordered container
-    using DeadlineQueue = std::map<std::chrono::steady_clock::time_point, QueueIterator>;
-
-
-    mutable std::shared_mutex rwlock;
-    Queue queue;
-
-    /// This is needed only for using inside cleanup() function and correct signaling about shutdown
-    mutable std::mutex cleanup_mutex;
-    mutable std::condition_variable cleanup_can_run;
+    using Queue = std::map<std::chrono::steady_clock::time_point, Container>;
+    using QueueIterator = Queue::iterator;
+    using QueueIteratorByKey = std::unordered_map<InsertQuery, QueueIterator, InsertQuery::Hash>;
 
-    mutable std::mutex deadline_mutex;
+    mutable std::mutex queue_mutex;
     mutable std::condition_variable are_tasks_available;
-    DeadlineQueue deadline_queue;
 
-    using QueryIdToEntry = std::unordered_map<String, InsertData::EntryPtr>;
-    mutable std::mutex currently_processing_mutex;
-    QueryIdToEntry currently_processing_queries;
+    Queue queue;
+    QueueIteratorByKey queue_iterators;
```

Comment:
It might be helpfull to add ```TSA_GUARDED_BY(queue_mutex)``` annotation to these two fields

Reply:
Looks like is doesn't work well with `unique_lock` and condition variable.